### PR TITLE
Add MemoryMonitor component and actions

### PR DESCRIPTION
**Pull request by knelli2**
## Proposed changes

Adds a singleton parallel component responsible for monitoring the memory of each parallel component. Also adds simple actions for each type of parallel component that will either compute or process the size of the local branch of a parallel component.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Parallel/MemoryMonitor/Tags.hpp
Comment (nilsdeppe) [inline src/Parallel/MemoryMonitor/Tags.hpp]: A design decision that's not clear to me is why you opted to have the component be a template parameter instead of using a map with a string as the key. Then you only need one tag in the DataBox instead of many, which simplifies the initialization action quite a bit. You actually wouldn't even need an initialization action anymore, you could trivially use the generic `AddSimpleTags` one. I think you can do that already.

File: src/ParallelAlgorithms/Actions/MonitorMemory.hpp
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Actions/MonitorMemory.hpp]: move to previous commit?

File: src/ParallelAlgorithms/Actions/MonitorMemory.hpp
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Actions/MonitorMemory.hpp]: I don't really love one file with so many different actions in it. My impression is the actions are all meant to be used in different places (maybe one or two in the same place). I think it's clear add either a subdirectory `MemoryMonitor` with several different files, or maybe something like `ProcessSingletonMemory.hpp`?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
I have not gone through all the code in detail, I have a couple higher level design questions before I do that. If it's better to talk on a call, just let me know!

File: src/Parallel/MemoryMonitor/Tags.hpp
Comment (knelli2) [reply src/Parallel/MemoryMonitor/Tags.hpp]: When I was writing this, it wasn't clear which components would actually use their tag in the memory monitor and which wouldn't. We always have access to the component we are sizing so it made sense to just template the tag on the component and then grab the tag corresponding to the component when I needed it.

However, with this design, the tags are only really needed for group and nodegroup components so a map of strings probably makes more sense.

File: src/ParallelAlgorithms/Actions/MonitorMemory.hpp
Comment (knelli2) [reply src/ParallelAlgorithms/Actions/MonitorMemory.hpp]: Besides tests, all these actions are going to be using in a `MonitorMemory` event I'll add in a future PR. `ContributeMemoryData` won't be called directly by the event but indirectly by the `ProcessGroup` and `ProcessNodegroup` actions.

That being said, this file is a bit crowded with different things and may be confusing. I can split it up into 4 files in a `MemoryMonitor` subdirectory:
- `ContributeMemoryData.hpp`
- `ProcessGroups.hpp` (has both group and nodegroup since they are so similar)
- `ProcessArray.hpp`
- `ProcessSingleton.hpp`

File: src/Parallel/MemoryMonitor/Tags.hpp
Comment (nilsdeppe) [reply src/Parallel/MemoryMonitor/Tags.hpp]: Okay, is it reasonable to make that change now? Probably before doing so it's worth thinking carefully if there will be any downsides. Nothing jumps out at me since we don't really need the compile time info in the tag, it's basically just used for naming things, and so we can just do the naming dynamically

File: src/ParallelAlgorithms/Actions/MonitorMemory.hpp
Comment (nilsdeppe) [reply src/ParallelAlgorithms/Actions/MonitorMemory.hpp]: Okay, that sounds good to me if that sounds like the right thing to you. You know all the followup PRs :)

File: src/Parallel/MemoryMonitor/Tags.hpp
Comment (knelli2) [reply src/Parallel/MemoryMonitor/Tags.hpp]: I don't see a downside either. It's basically just making the name a string at runtime vs a template at compile time. Nothing else

**Comment by knelli2** [comment]
@nilsdeppe Made those changes and shuffled a couple things around, but the bulk of the code stayed the same.

File: src/Parallel/MemoryMonitor/Tags.hpp
Comment (nilsdeppe) [inline src/Parallel/MemoryMonitor/Tags.hpp:30]: Do you want to make this a code block using triple ticks on the line before and after? Up to you :)

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp]: Make this a `REQUIRE` since the rest of the test won't give reasonable results if this doesn't pass?

File: src/Parallel/MemoryMonitor/MemoryMonitor.hpp
Comment (nilsdeppe) [inline src/Parallel/MemoryMonitor/MemoryMonitor.hpp:35]: can you document the `mem_monitor` namespace in this file? Not the `detail` one above, that way it'll show up in the dox, as well as everything inside of it. I would argue for putting the component in that namespace too. No reason all components have to be in the global namespace. Then I'd put the `mem_monitor` in the parallel doxygen group, but not the component itself

File: src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp:57]: What do you think of changing the name to `ContributeMemoryConsumption` or `MemoryUsed`? Something more specific than `Data`? Feel free to decline or defer to a PR at the end that does a global rename. I realize renames can be rather annoying in the middle of a series of PRs or commits :)

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp]: `const gsl`

File: src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp]: `const gsl`

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp:192]: we avoid disk IO in these types of tests by having a mock writing action that just records what was meant to be written and then checking that. Disk IO is slow, and just increases likelihood of bugs in general. Eg line 46 of `tests/Unit/Helpers/ParallelAlgorithms/LinearSolver/ResidualMonitorActionsTestHelpers.hpp`

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp]: pass runner as `gsl::not_null`

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp]: I'm a bit confused, would that mean you want the check _after_ the action?

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp]: isn't this closing a function?

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp]: `const gsl`

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp:487]: maybe add a comment here `// First just test `ContributeMemoryData` (last argument false), then test...`?

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp]: `const gsl`

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
overall LGTM, went through the code in detail this time.

File: src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp
Comment (knelli2) [reply src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp:57]: I think I'll wait until all the memory monitor stuff is in just so rebasing goes smoother.

How about `ContributeComponentSize`?

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp
Comment (knelli2) [reply tests/Unit/Parallel/Test_MemoryMonitor.cpp]: So before the first action is invoked on the memory monitor, the map for that time should be empty. And after the last action, the map should also be empty (it gets cleaned up). So if I move the check after the action, I'd have to do something like

```
if (i != num_branches-1) {
  CHECK(...);
}
```

to avoid an access error after the last action. So, it's just a matter of where you want to put the skip.

File: src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp
Comment (nilsdeppe) [reply src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp:57]: Yep, both sounds good!

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp
Comment (nilsdeppe) [reply tests/Unit/Parallel/Test_MemoryMonitor.cpp]: Ah okay, could you maybe rephrase the comment to include that detail? That is, `The map is empty both before and after the last call, so we need to put the CHECK inside an if clause regardless`?

**Comment by knelli2** [comment]
@nilsdeppe Rebased on develop and added fixups.

Re: the disk IO comment. I have opened a PR (#3969) to address this generally, and not just for this PR. Since this PR has multiple follow ups that depend on it, if it's ok with you, I'd like to forgo the changes to the test for now so we don't have to wait for #3969. Once #3969 is merged, I can open a follow-up to this PR that changes the test. Does that sound alright?

**Review by nilsdeppe — COMMENTED** [review]
LGTM, and I'm happy leaving the H5 refactor until later!

**Review by nilsdeppe — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
@kidder said he would like to review this, so waiting on him

File: src/Parallel/MemoryMonitor/MemoryMonitor.hpp
Comment (kidder) [inline src/Parallel/MemoryMonitor/MemoryMonitor.hpp:26]: this seems (not your fault) silly...
Not for this PR, but I think `AddSimpleTags` should just take a list of tags which are default initialized (which is already done automatically by the DataBox).  If you want a simple tag not default initialized you should just call an initialization action or `MutateApply` with a mutator.  Then we don't need this detail struct

File: src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp
Comment (kidder) [inline src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp]: MB -> megabytes

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp
Comment (kidder) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp]: remove this test because (as far as I can tell) it doesn't test anything except the mocking framework...

File: src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp
Comment (kidder) [inline src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp]: you don't need to pass the second argument to get a default constructed object

File: src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp
Comment (kidder) [inline src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp]: remove the if-else block, and just call `memory_holder.try_emplace(time);` followed by line 87

File: src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp
Comment (kidder) [inline src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp]: that -> than

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp
Comment (kidder) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp]: and -> an

File: src/ParallelAlgorithms/Actions/MemoryMonitor/ProcessGroups.hpp
Comment (kidder) [inline src/ParallelAlgorithms/Actions/MemoryMonitor/ProcessGroups.hpp]: isn't this just `array_index` (and thus can't we just have one action that works on both groups and nodegroups)?

**Review by kidder — CHANGES_REQUESTED** [review]

File: src/ParallelAlgorithms/Actions/MemoryMonitor/ProcessGroups.hpp
Comment (knelli2) [reply src/ParallelAlgorithms/Actions/MemoryMonitor/ProcessGroups.hpp]: You're right. `ContribtueMemoryData` expects an `int` for the node/proc so I just `static_assert`ed that the array index was an `int` inside `ProcessGroups`. The only component that doesn't have `int` for the array index is `Array` (and you shouldn't be running this action on an `Array`) so we would probably have been good without it, but I added it just in case.

**Review by kidder — COMMENTED** [review]
you have my permission to squash; @nilsdeppe did you want to look at the fixups?

**Comment by nilsdeppe** [comment]
I'm happy with the fixups and for you to squash and rebase

**Comment by knelli2** [comment]
@kidder @nilsdeppe Squashed and rebased!

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**
